node/src/components/contract_runtime/rewards.rs

```diff
@@ -169,203 +278,151 @@ impl ErasInfo {
+            effect_builder,
+            chainspec.protocol_version(),
+            chainspec.core_config.signature_rewards_max_delay,
+            executable_block,
+        )
+        .await?;
 
-        Ok(())
+        rewards_for_era(rewards_info, current_era_id, &chainspec.core_config)
     }
+}
 
+pub(crate) fn rewards_for_era(
+    rewards_info: RewardsInfo,
+    current_era_id: EraId,
+    core_config: &CoreConfig,
+) -> Result<BTreeMap<PublicKey, U512>, RewardsError> {
     fn to_ratio_u512(ratio: Ratio<u64>) -> Ratio<U512> {
         Ratio::new(U512::from(*ratio.numer()), U512::from(*ratio.denom()))
     }
 
-    // Special case: genesis block does not yield any reward, because there is no block producer,
-    // and no previous blocks whose signatures are to be rewarded:
-    if era_id.is_genesis() {
-        return Ok(BTreeMap::new());
-    }
+    let mut full_reward_for_validators: BTreeMap<_, _> = rewards_info
+        .validator_keys(current_era_id)?
+        .map(|key| (key, Ratio::new(U512::zero(), U512::one())))
+        .collect();
 
-    // All the blocks that may appear as a signed block. They are collected upfront, so that we
-    // don't have to worry about doing it one by one later.
-    //
-    // They are sorted from the oldest to the newest:
-    let cited_blocks = {
-        let cited_block_height_start = {
-            let previous_era_id = era_id.saturating_sub(1);
-            let previous_era_switch_block_header = effect_builder
-                .get_switch_block_header_by_era_id_from_storage(previous_era_id)
-                .await
-                .ok_or(RewardsError::MissingSwitchBlock(previous_era_id))?;
+    let mut increase_value_for_key =
+        |key: PublicKey, value: Ratio<U512>| -> Result<(), RewardsError> {
+            match full_reward_for_validators.entry(key) {
+                std::collections::btree_map::Entry::Vacant(entry) => {
+                    entry.insert(value);
+                }
+                std::collections::btree_map::Entry::Occupied(mut entry) => {
+                    let new_value = (*entry.get())
+                        .checked_add(&value)
+                        .ok_or(RewardsError::ArithmeticOverflow)?;
+                    *entry.get_mut() = new_value;
+                }
+            }
 
-            // Here we do not substract 1, because we want one block more:
-            previous_era_switch_block_header
-                .height()
-                .saturating_sub(chainspec.core_config.signature_rewards_max_delay)
+            Ok(())
         };
-        let range_to_fetch = cited_block_height_start..switch_block_height;
 
-        let result = collect_past_blocks_batched(effect_builder, range_to_fetch.clone()).await;
+    // Rules out a special case: genesis block does not yield any reward,
+    // because there is no block producer, and no previous blocks whose
+    // signatures are to be rewarded:
+    if current_era_id.is_genesis() == false {
```

Comment:
Can this ever be called with the genesis era? `fetch_data_and_calculate_rewards_for_era` already checks for that case.

Reply:
That's right, I replaced it with a debug assertion.